tests/dom/customElement.ts

```diff
@@ -84,3 +85,94 @@ test('types text inside custom element', async () => {
     input[value="Sup"] - keyup: p
   `)
 })
+
+const template = document.createElement('template')
+template.innerHTML = `
+  <input>
+`
+
+class ShadowInput extends HTMLElement {
+  constructor() {
+    super()
+
+    this.attachShadow({mode: 'open', delegatesFocus: true})
+    if (this.shadowRoot) {
+      this.shadowRoot.appendChild(template.content.cloneNode(true))
+    }
+  }
+
+  public get value(): string {
+    return this.shadowRoot?.querySelector('input')?.value ?? ''
+  }
+}
+window.customElements.define('shadow-input', ShadowInput)
+
+test('Render open shadow DOM element - type', async () => {
+  const {element, user} = setup('<shadow-input></shadow-input>', {
+    focus: false,
+  })
+
+  const inputElement = element.shadowRoot?.querySelector('input')
+  if (inputElement) {
+    await user.click(inputElement)
+  }
+  await user.keyboard('test')
+
+  expect((element as ShadowInput).value).toEqual('test')
+})
+
+test('Render open shadow DOM element - copy', async () => {
+  const {element, user} = setup('<shadow-input></shadow-input>', {
+    focus: false,
+  })
+
+  const inputElement = element.shadowRoot?.querySelector('input')
+  if (inputElement) {
+    await user.click(inputElement)
+  }
+  await user.keyboard('test')
+
+  if (inputElement) {
+    setUISelection(inputElement, {anchorOffset: 0, focusOffset: 4})
+  }
```

Comment:
The tests should make clear which feature is tested and be narrowed down to this - i.e. no other user-event API being used. Otherwise it might be hard for the next developer to determine what is tested, what lacks tests and which tests are redundant. See the comment in this particular file above.

Reply:
So I added support for `selection` and `focus` in `setup()`, so this is no longer necessary.